File: Shortest Subarray with Sum at Least K.cpp
class Solution {
public:
    int shortestSubarray(vector<int>& nums, int k) {
        int n = nums.size();
        int answer = INT_MAX;
        deque<int> q;
        vector<long> sums(n+1);
        for(int i=0; i<nums.size(); i++){
            sums[i+1] = sums[i]+ (long)nums[i];
        }
        for(int j = 0; j<sums.size(); j++){
            while(q.size() && sums[j]<=sums[q.back()])
                q.pop_back();
            while(q.size() && sums[j] >=sums[q.front()]+k){
                answer = min(answer, j-q.front());
                q.pop_front();
            }
            q.push_back(j);
        }
        if (answer == INT_MAX)
            return -1; 
        return answer;
    }
};

File: Shortest Subarray with Sum at Least K.py
class Solution:
    def shortestSubarray(self, nums: List[int], k: int) -> int:
        mono = deque()
        shortest = len(nums)+1
        prefix = [0]*(len(nums)+1)

        for i in range(1,len(prefix)):
            prefix[i] = prefix[i-1] + nums[i-1]

        for right in range(len(prefix)):
            while len(mono)!=0 and prefix[right] - prefix[mono[0]] >= k:
                shortest = min(shortest, right-mono.popleft())     
            while len(mono) != 0 and prefix[mono[-1]] > prefix[right]:
                mono.pop()
            mono.append(right)
        if shortest == len(nums)+1:
            return -1
        return shortest
